### PR TITLE
Added platform user apis

### PR DESCRIPTION
**Pull request by aman-tunerlabs**

### CODE REVIEW

File: controllers/v1/solutions.js [hunk 1]
```diff
@@ -813,4 +813,61 @@ module.exports = class Solutions extends Abstract {
...
+   /**
+   * List solutions by program id.
+   * @method
+   * @name listByProgramId
+   * @param {Object} req - requested data.
+   * @param {String} req.params._id - program id .
+   * @returns {Array}
+   */
+
+    async listByProgramId(req) {
```
Comment (aks30) [inline controllers/v1/solutions.js:852]: @aman-tunerlabs as discussed move this API in userExtension and validate program access for the user.

File: controllers/v1/user-extension.js [hunk 2]
```diff
@@ -482,4 +482,117 @@ module.exports = class UserExtension extends Abstract {
...
+
+
+  /**
+   * List of user platform roles
+   * @method
+   * @name platformRoles
+   * @returns {Array} List of user platform roles
+   */
+
+   platformRoles(req) {
```
Comment (aks30) [inline controllers/v1/user-extension.js:515]: @aman-tunerlabs - Let's discuss if we need this API for now or not.

File: controllers/v1/user-extension.js [hunk 3]
```diff
@@ -482,4 +482,117 @@ module.exports = class UserExtension extends Abstract {
...
+          errorObject: error
+        })
+      }
+
+
+    })
+  }
+   
+    /**
+  * @api {get} /kendra/api/v1/user-extension/programsByPlatformRoles?role=:role List of programs for platform user
```
Comment (aks30) [inline controllers/v1/user-extension.js:541]: @aman-tunerlabs - Make role optional

File: module/solutions/helper.js [hunk 4]
```diff
@@ -1470,6 +1470,52 @@ module.exports = class SolutionsHelper {
...
 
+    /**
+   * List solutions.
+   * @method  
+   * @name listByProgramId
+   * @param {String} programId - Program id.
+   * @returns {JSON} List of solutions.
+   */
+  
+  static listByProgramId(programId) {
```
Comment (aks30) [inline module/solutions/helper.js]: @aman-tunerlabs Not required as discussed.

File: module/solutions/validator/v1.js [hunk 5]
```diff
@@ -49,6 +49,9 @@ module.exports = (req) => {
         },
         targetedEntity : function () {
             req.checkParams("_id").exists().withMessage("required solution id");
+        },
+        listByProgramId: function () {
```
Comment (aks30) [inline module/solutions/validator/v1.js]: @aman-tunerlabs Not required as discussed.

File: module/user-extension/helper.js [hunk 6]
```diff
@@ -1024,6 +1025,108 @@ static updateUserRolesInEntitiesElasticSearch(userId = "", userRoles = []) {
...
+
+   /**
+   * List of user platform roles
+   * @method
+   * @name platformRoles
+   * @param {String} userId - Logged in user id.
+   * @returns {Array} 
+   */
+
+  static platformRoles(userId) {
```
Comment (aks30) [inline module/user-extension/helper.js:1038]: @aman-tunerlabs - Maybe removed after discussion.

File: module/user-extension/helper.js [hunk 7]
```diff
@@ -1024,6 +1025,108 @@ static updateUserRolesInEntitiesElasticSearch(userId = "", userRoles = []) {
...
+    static programsByPlatformRoles(userId,role) {
+        return new Promise(async (resolve, reject) => {
+            try {
+                
+                const userInformation = await this.userExtensionDocument
+                (
+                    { 
+                        userId: userId,
+                        status: constants.common.ACTIVE,
+                        "platformRoles.code": role
```
Comment (aks30) [inline module/user-extension/helper.js]: @aman-tunerlabs This query parameter will be optional

File: module/user-extension/validator/v1.js [hunk 8]
```diff
@@ -11,6 +11,9 @@ module.exports = (req) => {
         updateProfileRoles : function () {
             req.checkBody('stateId').exists().withMessage("State id is required");
             req.checkBody('roles').exists().withMessage("Roles data is required");
+        },
+        programsByPlatformRoles: function () {
```
Comment (aks30) [inline module/user-extension/validator/v1.js]: @aman-tunerlabs As discussed it can be optional

File: controllers/v1/solutions.js (hunk 1 above)
Comment (aman-tunerlabs) [reply controllers/v1/solutions.js:852]: @aks30  Resolved

File: module/solutions/helper.js (hunk 4 above)
Comment (aman-tunerlabs) [reply module/solutions/helper.js]: @aks30  Resolved

File: controllers/v1/user-extension.js [hunk 9]
```diff
@@ -482,4 +482,187 @@ module.exports = class UserExtension extends Abstract {
...
+          errorObject: error
+        })
+      }
+
+
+    })
+  }
+
+    /**
+  * @api {get} /kendra/api/v1/user-extension/solutionsByPlatformProgram/:programId?role=:role List of solutions for platform user program
```
Comment (aks30) [inline controllers/v1/user-extension.js]: @aman-tunerlabs change endpoint to /solutions/:programId

File: module/user-extension/validator/v1.js [hunk 10]
```diff
@@ -11,6 +11,9 @@ module.exports = (req) => {
         updateProfileRoles : function () {
             req.checkBody('stateId').exists().withMessage("State id is required");
             req.checkBody('roles').exists().withMessage("Roles data is required");
+        },
+        solutionsByPlatformProgram: function () {
```
Comment (aks30) [inline module/user-extension/validator/v1.js]: @aman-tunerlabs - changed endpoint to solutions

File: module/solutions/validator/v1.js [hunk 11]
```diff
@@ -49,6 +49,7 @@ module.exports = (req) => {
         },
         targetedEntity : function () {
             req.checkParams("_id").exists().withMessage("required solution id");
+            req.checkQuery("role").exists().withMessage("required user role");
```
Comment (aks30) [inline module/solutions/validator/v1.js]: @aman-tunerlabs check if this is required.

**Pull request merged**